File: CustomUiComponents/app/src/main/java/com/jarvis/dragdropresearch/funwithshapes/FlashShape.java
package com.jarvis.dragdropresearch.funwithshapes;

import android.graphics.Color;

import com.jarvis.dragdropresearch.interpolators.AlphaInterpolator;
import com.jarvis.dragdropresearch.interpolators.ColorInterpolator;

import java.util.Random;

public abstract class FlashShape {
    private static final int[] COMPONENT_COLOR_POOL_DEFAULT =
            new int[] {Color.RED, Color.GREEN};

    private Type mType;

    private int mXOffset;

    private int mYOffset;

    private ColorInterpolator mColorInterpolator;

    private AlphaInterpolator mAlphaInterpolator;

    private boolean mAllowMulticoloredComponents;

    private int[] mComponentColors;

    private int[] mComponentColorPool = COMPONENT_COLOR_POOL_DEFAULT;

    FlashShape(Type type) {
        mType = type;
        mComponentColors = mComponentColorPool;
    }

    public Type getType() {
        return mType;
    }

    public int getXOffset() {
        return mXOffset;
    }

    public void setXOffset(int XOffset) {
        mXOffset = XOffset;
    }

    public int getYOffset() {
        return mYOffset;
    }

    public void setYOffset(int YOffset) {
        mYOffset = YOffset;
    }

    public int[] getComponentColors() {
        return mComponentColors;
    }

    /**
     * @param componentColors Array of colors to be used for coloring each segment. The segments will be
     * colored in order of componentColors[segmentIndex % componentColors.length] to account for the
     * total segment count being greater than the number of segment colors.
     */
    public void setComponentColors(int[] componentColors) {
        mComponentColors = componentColors;
    }

    public int[] getComponentColorPool() {
        return mComponentColorPool;
    }

    /**
     * @return Maximum number of components that this shape can be composed of.
     */
    abstract public int getMaxComponents();

    public boolean allowMultiColoredComponents() {
        return mAllowMulticoloredComponents;
    }

    /**
     * @param allowMulticoloredComponents Set to true if individual segment colors should be respected.
     * If false,the default color used to draw the spiral will be used for all segments.
     */
    public void setAllowMulticoloredComponents(boolean allowMulticoloredComponents) {
        mAllowMulticoloredComponents = allowMulticoloredComponents;
    }

    /**
     * @param componentColorPool Array of colors to be used in conjunction with
     * {@link SpiralShape#generateRandomComponentColors()}. If null or the length is
     * equal to 0, the default color pool will be used.
     *
     * @see #COMPONENT_COLOR_POOL_DEFAULT
     */
    public void setComponentColorPool(int[] componentColorPool) {
        if (componentColorPool != null && componentColorPool.length > 0) {
            mComponentColorPool = componentColorPool;
        } else {
            mComponentColorPool = COMPONENT_COLOR_POOL_DEFAULT;
        }
    }

    /**
     * Generate a random array of colors of size.
     * This array will be used to color the segments, with array[i] coloring segment[i].
     */
    public void generateRandomComponentColors() {
        mComponentColors = new int[getMaxComponents()];

        final Random random = new Random(System.currentTimeMillis());
        for (int i = 0; i < mComponentColors.length; i++) {
            mComponentColors[i] = mComponentColorPool[random.nextInt(300) %
                    mComponentColorPool.length];
        }
    }

    public ColorInterpolator getColorInterpolator() {
        return mColorInterpolator;
    }

    public void setColorInterpolator(
            ColorInterpolator colorInterpolator) {
        mColorInterpolator = colorInterpolator;
    }

    public AlphaInterpolator getAlphaInterpolator() {
        return mAlphaInterpolator;
    }

    public void setAlphaInterpolator(
            AlphaInterpolator alphaInterpolator) {
        mAlphaInterpolator = alphaInterpolator;
    }

    public enum Type {
        ARC,
        RECTANGLE,
        SPIRAL,
        STAR,
        TRIANGLE
    }
}

File: CustomUiComponents/app/src/main/java/com/jarvis/dragdropresearch/funwithshapes/FlashShape.kt
package com.jarvis.dragdropresearch.funwithshapes

import com.jarvis.dragdropresearch.funwithshapes.enums.ColorPalette
import com.jarvis.dragdropresearch.interpolators.AlphaInterpolator
import com.jarvis.dragdropresearch.interpolators.ColorInterpolator
import java.util.*

abstract class FlashShape(type: Type) {

    private val mType = type

    private var mXOffset: Int = 0

    private var mYOffset: Int = 0

    private var mColorInterpolator: ColorInterpolator? = null

    private var mAlphaInterpolator: AlphaInterpolator? = null

    private var mAllowMulticoloredComponents: Boolean = false

    private var mComponentColorPool = COMPONENT_COLOR_POOL_DEFAULT

    private var mComponentColors: IntArray? = null

    init {
        mComponentColors = mComponentColorPool
    }

    fun getType(): Type {
        return mType
    }

    fun getXOffset(): Int {
        return mXOffset
    }

    fun setXOffset(XOffset: Int) {
        mXOffset = XOffset
    }

    fun getYOffset(): Int {
        return mYOffset
    }

    fun setYOffset(YOffset: Int) {
        mYOffset = YOffset
    }

    fun getComponentColors(): IntArray? {
        return mComponentColors
    }

    fun getColorInterpolator(): ColorInterpolator? {
        return mColorInterpolator
    }

    fun setColorInterpolator(colorInterpolator: ColorInterpolator) {
        mColorInterpolator = colorInterpolator
    }

    fun getAlphaInterpolator(): AlphaInterpolator? {
        return mAlphaInterpolator
    }

    fun setAlphaInterpolator(alphaInterpolator: AlphaInterpolator) {
        mAlphaInterpolator = alphaInterpolator
    }

    /**
     * @param componentColors Array of colors to be used for coloring each segment. The segments will be
     * colored in order of componentColors[segmentIndex % componentColors.length] to account for the
     * total segment count being greater than the number of segment colors.
     */
    fun setComponentColors(componentColors: IntArray) {
        mComponentColors = componentColors
    }

    fun getComponentColorPool(): IntArray {
        return mComponentColorPool
    }

    /**
     * @return Maximum number of components that this shape can be composed of.
     */
    abstract fun getMaxComponents(): Int

    fun allowMultiColoredComponents(): Boolean {
        return mAllowMulticoloredComponents
    }

    /**
     * @param allowMulticoloredComponents Set to true if individual segment colors should be respected.
     * If false,the default color used to draw the spiral will be used for all segments.
     */
    fun setAllowMulticoloredComponents(allowMulticoloredComponents: Boolean) {
        mAllowMulticoloredComponents = allowMulticoloredComponents
    }

    /**
     * @param componentColorPool Array of colors to be used in conjunction with
     * [SpiralShape.generateRandomComponentColors]. If null or the length is
     * equal to 0, the default color pool will be used.
     *
     * @see .COMPONENT_COLOR_POOL_DEFAULT
     */
    fun setComponentColorPool(componentColorPool: IntArray?) {
        if (componentColorPool != null && componentColorPool.isNotEmpty()) {
            mComponentColorPool = componentColorPool
        } else {
            mComponentColorPool = COMPONENT_COLOR_POOL_DEFAULT
        }
    }

    /**
     * Generate a random array of colors of size.
     * This array will be used to color the segments, with array[i] coloring segment[i].
     */
    fun generateRandomComponentColors() {
        val componentColors = IntArray(getMaxComponents())

        val random = Random(System.currentTimeMillis())

        for (i in componentColors.indices) {
            componentColors[i] =
                mComponentColorPool[random.nextInt(300) % mComponentColorPool.size]
        }

        mComponentColors = componentColors
    }

    companion object {
        private val COMPONENT_COLOR_POOL_DEFAULT = ColorPalette.ONE.values
    }

    enum class Type {
        ARC,
        RECTANGLE,
        SPIRAL,
        STAR,
        TRIANGLE
    }
}
